Legacy device API handles Prod observables

astralcai opened the pull request:
**Context:**
The legacy device API has special handling of `Tensor` and `Sum` as observables, but `Prod` is not covered by that.

**Description of the Change:**
- `Prod` observables are expanded if they simplify to a `Sum`
- In `check_validity`, the operands of `Prod` observables are checked.

**Possible Drawbacks:**
`simplify` is called on `Prod` observables many times, which might be inefficient

**Related Shortcut Story:**
[sc-60584]

astralcai commented on tests/test_device.py:
```diff
@@ -30,36 +30,6 @@
 # pylint: disable=abstract-class-instantiated, no-self-use, redefined-outer-name, invalid-name, missing-function-docstring
 
 
-@pytest.fixture(scope="function")
-def mock_device_with_operations(monkeypatch):
-    """A function to create a mock device with non-empty operations"""
-    with monkeypatch.context() as m:
-        m.setattr(Device, "__abstractmethods__", frozenset())
-        m.setattr(Device, "operations", mock_device_paulis)
-        m.setattr(Device, "observables", mock_device_paulis)
-        m.setattr(Device, "short_name", "MockDevice")
-
-        def get_device(wires=1):
-            return Device(wires=wires)
-
-        yield get_device
-
-
-@pytest.fixture(scope="function")
-def mock_device_with_observables(monkeypatch):
-    """A function to create a mock device with non-empty observables"""
-    with monkeypatch.context() as m:
-        m.setattr(Device, "__abstractmethods__", frozenset())
-        m.setattr(Device, "operations", mock_device_paulis)
-        m.setattr(Device, "observables", mock_device_paulis)
-        m.setattr(Device, "short_name", "MockDevice")
-
-        def get_device(wires=1):
-            return Device(wires=wires)
-
-        yield get_device
-
-
```
They are exact duplicates of another mock device.

astralcai commented on tests/test_device.py:
```diff
@@ -203,19 +173,15 @@ def get_device(wires=1):
+        def get_device(wires=1):
             return Device(wires=wires)
 
         yield get_device
```
Another mock device that is a duplicate.

astralcai commented on tests/test_device.py:
```diff
@@ -203,19 +173,15 @@ def get_device(wires=1):
 
 @pytest.fixture(scope="function")
-def mock_device_arbitrary_wires(monkeypatch):
+def mock_device_supporting_prod(monkeypatch):
```
This is an added device to test `Prod` support

Qottmann commented on pennylane/_device.py:
```diff
@@ -759,7 +758,12 @@ def batch_transform(self, circuit: QuantumTape):
         expval_sum_in_obs = any(
-            isinstance(m.obs, Sum) and isinstance(m, ExpectationMP) for m in circuit.measurements
+            (
+                isinstance(m.obs, Sum)
+                or (isinstance(m.obs, Prod) and isinstance(m.obs.simplify(), Sum))
```
At this point why not include SProd as well?

Qottmann replied:
Couldnt the same be said about a sum?

That being said, not sure this is necessary here :)

Qottmann commented on pennylane/_device.py:
```diff
@@ -1007,6 +1011,21 @@ def check_validity(self, queue, observables):
+
+                supports_prod = self.supports_observable(o.name)
+                if not supports_prod:
+                    raise DeviceError(f"Observable Prod not supported on device {self.short_name}")
```
```suggestion
                    raise DeviceError(f"Observable {o} of type {type(o)} not supported on device {self.short_name}")
```

Qottmann commented on pennylane/_device.py:
```diff
@@ -1007,6 +1011,21 @@ def check_validity(self, queue, observables):
+                    for i in o.simplify().operands:
+                        if not self.supports_observable(i.name):
+                            raise DeviceError(
+                                f"Observable {i.name} not supported on device {self.short_name}"
```
```suggestion
                                f"Observable {i} of type {type(i)} not supported on device {self.short_name}"
```

astralcai replied:
I felt that the original error message was quite clear so I didn't change anything.

Qottmann commented on pennylane/devices/tests/test_measurements.py:
```diff
@@ -155,8 +155,10 @@ class TestHamiltonianSupport:
     """Separate test to ensure that the device can differentiate Hamiltonian observables."""
 
     @pytest.mark.parametrize("ham_constructor", [qml.ops.Hamiltonian, qml.ops.LinearCombination])
+    @pytest.mark.filterwarnings("ignore::pennylane.PennyLaneDeprecationWarning")
```
how come we need this now? 🤔 and what is raising the deprecation warning?

astralcai replied:
The test case is explicitly constructing `qml.ops.Hamiltonian`, see the `ham_constructor` parametrization.

Qottmann replied:
But how come this has passed CI before this 😬 ? Shouldnt we get a fail from qml.PennylaneDeprecationWarnings?

astralcai replied:
This is not part of PennyLane's CI. It is part of the device tests, and is used by the plugin test matrix. I don't think the plugin test matrix treats `PennyLaneDeprecationWarning` as failures, but we do see it in the warnings summary: https://github.com/PennyLaneAI/plugin-test-matrix/actions/runs/8624006838/job/23638246184

Qottmann replied:
Ah that explains it, thanks!

Qottmann commented on tests/test_device.py:
```diff
@@ -340,6 +306,33 @@ def test_check_validity_on_valid_queue(self, mock_device_supporting_paulis):
+        dev.check_validity(queue, observables)
+
+        unsupported_nested_observables = [
+            qml.expval(qml.PauliZ(0) @ (qml.PauliX(1) @ qml.PauliY(2)))
```
Perhaps I am standing on the hose but what is different here? 🤯

astralcai replied:
The mock device I made did not support `PauliY`, so this is to test that a `Prod` of `PauliZ` and another `Prod` is handled properly, that is, the whole thing unwraps, and support for each operand is checked.

Qottmann replied:
Aaaaah I missed the final op being a *Y*, thanks!

Qottmann reviewed (commented):
Great work! Thanks @astralcai !

I left mostly questions for now :)

Concerning the renaming of the mock devices in the test I dont have an opinion and would refer to someone from core to judge whether changing all of them is a good idea or not.

astralcai commented:
> Concerning the renaming of the mock devices in the test I dont have an opinion and would refer to someone from core to judge whether changing all of them is a good idea or not.

@Qottmann The renaming was because multiple test cases were using different mock devices that were exact duplicates of one another. I removed the duplicates, and changed all these test cases to use just the one that was left.

lillian542 commented on pennylane/transforms/hamiltonian_expand.py:
```diff
@@ -341,6 +341,8 @@ def sum_expand(tape: QuantumTape, group: bool = True) -> (Sequence[QuantumTape],
     for idx, m in enumerate(tape.measurements):
         obs = m.obs
+        if isinstance(obs, Prod) and isinstance(m, ExpectationMP):
+            obs = obs.simplify()
```
What breaks without this? Can we add the corresponding test that would fail without this to the test suite?

Qottmann replied:
Was also wondering this 🤔

Qottmann replied:
I would only keep this if there is a very good reason for it, as simplification can be rather expensive

lillian542 reviewed (commented):
Looks good @astralcai! Left a couple of small comments about tests.

astralcai commented on tests/test_device.py:
```diff
@@ -340,6 +306,33 @@ def test_check_validity_on_valid_queue(self, mock_device_supporting_paulis):
         # Raises an error if queue or observables are invalid
         dev.check_validity(queue, observables)
 
+    def test_check_validity_containing_prod(self, mock_device_supporting_prod):
```
Will do.

Qottmann reviewed (commented):
Thanks @astralcai 

My questions are mostly answered and I am overall happy with the PR.
The only thing left to discuss is the simplify. Why do we need it?

astralcai commented:
> Thanks @astralcai
> 
> My questions are mostly answered and I am overall happy with the PR. The only thing left to discuss is the simplify. Why do we need it?

@Qottmann The purpose of this is to expand a nested `Prod` out into either a `Prod` or a `Sum`. For example if the observable was `qml.X(0) @ qml.Y(1) @ qml.Z(2)`, we get a `Prod` with two operands, a `Prod` and a `Z`, or if the observable was `qml.X(0) @ (qml.Y(1) + qml.Z(2))`, we would also want to flatten out the nested structure. I know `simplify` does this, but do you know if there is another less expensive alternative?

Qottmann reviewed (approved):
Looks good to me for now, happy to move on and revisit later for a better solution with simplify 👍 

Thanks @astralcai great work!

lillian542 reviewed (approved):
What @Qottmann said 🚀
